### PR TITLE
[RISCV] Align stack size down to a multiple of 16 before using cm.push/pop.

### PR DESCRIPTION
**Pull request by topperc**
This an alternative to #84935 to fix the miscompile, but not be optimal.

The immediate for cm.push/pop must be a multiple of 16. For RVE, it might not be. It's not easy to increase the stack size without messing up cfa directives and maybe other things.

This patch rounds the stack size down to a multiple of 16 before clamping it to 48. This causes an extra addi to be emitted to handle the remainder.

Once this commited, I can commit #84989 to add verification for these instructions being generated with valid offsets.

### CODE REVIEW

**Comment by topperc** [comment]
Ping

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -554,8 +554,10 @@ void RISCVFrameLowering::emitPrologue(MachineFunction &MF,
   if (RVFI->isPushable(MF) && FirstFrameSetup != MBB.end() &&
       FirstFrameSetup->getOpcode() == RISCV::CM_PUSH) {
     // Use available stack adjustment in push instruction to allocate additional
-    // stack space.
-    uint64_t Spimm = std::min(StackSize, (uint64_t)48);
+    // stack space. Align the stack size down to a multiple of 16. This is
+    // needed for RVE.
+    // FIXME: Can we increase the stack size to a multiple of 16 insead?
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:559]: `insead`->`instead`

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -776,8 +778,10 @@ void RISCVFrameLowering::emitEpilogue(MachineFunction &MF,
   if (RVFI->isPushable(MF) && MBBI != MBB.end() &&
       MBBI->getOpcode() == RISCV::CM_POP) {
     // Use available stack adjustment in pop instruction to deallocate stack
-    // space.
-    uint64_t Spimm = std::min(StackSize, (uint64_t)48);
+    // space. Align the stack size down to a multiple of 16. This is needed for
+    // RVE.
+    // FIXME: Can we increase the stack size to a multiple of 16 insead?
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:783]: ditto.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 3]
```diff
@@ -554,8 +554,10 @@ void RISCVFrameLowering::emitPrologue(MachineFunction &MF,
   if (RVFI->isPushable(MF) && FirstFrameSetup != MBB.end() &&
       FirstFrameSetup->getOpcode() == RISCV::CM_PUSH) {
     // Use available stack adjustment in push instruction to allocate additional
-    // stack space.
-    uint64_t Spimm = std::min(StackSize, (uint64_t)48);
+    // stack space. Align the stack size down to a multiple of 16. This is
+    // needed for RVE.
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:558]: I may miss something here, why is this needed for RVE? Should all be 16-bytes aligned?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:558]: RVE stack alignment is XLen-bytes not 16-bytes

**Review by wangpc-pp — APPROVED** [review]
LGTM.
